coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go

```diff
@@ -29,7 +29,37 @@ var rLog = logger.NewLogger("coa.runtime")
 
 const (
 	entryCountPerList = 100
-	separator         = "*"
+	// separator should be disallowed characters in kubernetes custom resource names except for
+	// "." (some resource type contains it), "*" (stands for any string in regex), "_" (used in some prefix like sch_)
+	separator = "|"
+
+	// SetDefaultQuery is the lua script to set the default value
+	// 1. If the etag is not set, set always succeeds
+	// 2. If the etag is set, set succeeds only if
+	// 		1) the etag matches the current etag
+	// 		2) the etag "0" and the first-write flag is not set
+	setDefaultQuery = `
+	local etag = redis.pcall("HGET", KEYS[1], "version");
+	if type(etag) == "table" then
+	  redis.call("DEL", KEYS[1]);
+	end;
+	local fwr = redis.pcall("HGET", KEYS[1], "first-write");
+	if not etag or type(etag)=="table" or etag == "" or etag == ARGV[1] or (not fwr and ARGV[1] == "0") then
+	  redis.call("HSET", KEYS[1], "values", ARGV[2]);
+	  if ARGV[3] == "1" then
+	    redis.call("HSET", KEYS[1], "first-write", 1);
```

Comment:
It reads like the first-write field is actually used to distinguish between first write request from other requests? What if two clients sending the request at the same time?

Reply:
This logic is from the Dapr project and I think this is to guarantee all the further operations should use proper etag if any one is firstwrite. If you have different expectation, I can change it based on our need.